### PR TITLE
Feature/create arrears action diary entry

### PR DESCRIPTION
**Pull request by madetech-jeffpinkham**
API Endpoint for Creating action diary entries

### CODE REVIEW

File: LBHTenancyAPI/Services/CredentialsService.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+using AgreementService;
+
+namespace LBHTenancyAPI.Services
+{
+    public class CredentialsService : ICredentialsService
+    {
+        public string GetUhSourceSystem()
+        {
+            return "HackneyAPI";
```
[the remark this answers is not in the document]
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Services/CredentialsService.cs]: Yep agreed

[the remark this answers is not in the document]
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Services/CredentialsService.cs]: I need to do the ENV piece we discussed on Tuesday, to accomodate that.

[the remark this answers is not in the document]
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Services/CredentialsService.cs]: I can do that as part of this task or as a separate task

[the remark this answers is not in the document]
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Services/CredentialsService.cs]: I'll do the ENV piece now 👍

File: LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+using System.Threading.Tasks;
+using AgreementService;
+using LBHTenancyAPI.Interfaces;
+using LBHTenancyAPI.UseCases;
+
+namespace LBHTenancyAPI.Gateways
+{
+    public class ArrearsActionDiaryGateway : IArrearsActionDiaryGateway
+    {
+        private readonly IArrearsAgreementService _arrearsAgreementService;
```
[the remark this answers is not in the document]
Comment (StevenLeighton21) [reply LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs]: Should it be called actionDiaryService? Or does the service cover both creating action diary entries and agreements?

File: LBHTenancyAPI/Services/CredentialsService.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (StevenLeighton21) [reply LBHTenancyAPI/Services/CredentialsService.cs]: I think we should do that as part of this task, purely to avoid committing creds to the open source repo.

[the remark this answers is not in the document]
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Services/CredentialsService.cs]: YEAH! The repo is open source, isn't it.. Yep on it now.

File: LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs]: Yep agreed 👍

**Comment by madetech-jeffpinkham** [comment]
I've moved the configuration into the ECS ENV variables, and used EnvironmentVariables in the configuration. I deleted the tests that tested the configuration for Credentials Service as it's configuration and I found it a nightmare to test.

File: LBHTenancyAPI.sln [hunk 3]
```diff
@@ -31,4 +33,10 @@ Global
 		{B63ADA1B-A6C3-4DA7-B840-2E5ADFF6D95B}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{B63ADA1B-A6C3-4DA7-B840-2E5ADFF6D95B}.Release|Any CPU.ActiveCfg = Release|Any CPU
 	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
+	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
+		SolutionGuid = {7524071F-1952-4ABE-97E9-4E73D0F9BE47}
+	EndGlobalSection
 EndGlobal
```
Comment (zuren) [inline LBHTenancyAPI.sln]: Are these config changes actually necessary, or are they susceptible to churn every time a different dev pushes up?

File: LBHTenancyAPI.v3.ncrunchsolution [hunk 4]
```diff
@@ -0,0 +1,6 @@
+﻿<SolutionConfiguration>
+  <Settings>
+    <AllowParallelTestExecution>True</AllowParallelTestExecution>
+    <SolutionConfigured>True</SolutionConfigured>
+  </Settings>
+</SolutionConfiguration>
```
Comment (zuren) [inline LBHTenancyAPI.v3.ncrunchsolution]: Same here - possibly wise to gitignore? What's conventional?

File: LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿{
+  "ProviderId": "Microsoft.VisualStudio.ConnectedService.Wcf",
+  "Version": "15.0.20119.1312",
+  "ExtendedData": {
+    "Uri": "http://lbhuhwwebp01/civica-uh-gateway-dev/webservices/ArrearsAgreementService.svc",
```
Comment (zuren) [inline LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json]: Would consider this private, should be configured via the environment somehow IMO.

Comment (StevenLeighton21) [reply LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json]: That URI is on the live internet, no? It's accessible from outside prem? If so, I'd argue it's not private. If not, agree :)

File: LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using AgreementService;
+using LBHTenancyAPI.Extensions.Validation;
+using LBHTenancyAPI.UseCases.ArrearsActions;
+
+// For more information on enabling MVC for empty projects, visit https://go.microsoft.com/fwlink/?LinkID=397860
```
Comment (zuren) [inline LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: Probably unnecessary?

File: LBHTenancyAPI/Controllers/TenanciesController.cs [hunk 7]
```diff
@@ -58,6 +58,7 @@ public async Task<IActionResult> Get([FromQuery(Name = "tenancy_refs")] List<str
                 {"tenancies", tenancies}
             };
 
+
```
Comment (zuren) [inline LBHTenancyAPI/Controllers/TenanciesController.cs]: Unnecessary change?

File: LBHTenancyAPI/Dockerfile [hunk 8]
```diff
@@ -1,12 +1,13 @@
...
 
-FROM microsoft/aspnetcore-build:2.0 AS build
+FROM microsoft/dotnet:latest AS build
 WORKDIR /src
 COPY LBHTenancyAPI.sln ./
-COPY LBHTenancyAPI/LBHTenancyAPI.csproj LBHTenancyAPI/
-RUN dotnet restore -nowarn:msb3202,nu1503
+COPY LBHTenancyAPI/LBHTenancyAPI.csproj LBHTenancyAPI/
+RUN ls
+RUN dotnet restore LBHTenancyAPI/LBHTenancyAPI.csproj -nowarn:msb3202,nu1503
```
Comment (zuren) [inline LBHTenancyAPI/Dockerfile]: Looks like there's a debug "ls" left over in here. Are the changes to this file actually necessary now that the config has been fixed, or was this just experimenting?

File: LBHTenancyAPI/LBHTenancyAPI.csproj [hunk 9]
```diff
@@ -23,5 +28,36 @@
...
+  <!--<ItemGroup>
+    <WebReferences Include="Web References" />
+  </ItemGroup>-->
+  <!--<ItemGroup>
+    <WebReferenceUrl Include="http://lbhuhwwebp01/civica-uh-gateway-dev/webservices/ArrearsAgreementService.svc">
+      <UrlBehavior>Dynamic</UrlBehavior>
+      <RelPath>Web References\ArrearsAgreementService</RelPath>
+      <UpdateFromURL>http://lbhuhwwebp01/civica-uh-gateway-dev/webservices/ArrearsAgreementService.svc</UpdateFromURL>
+    </WebReferenceUrl>
+  </ItemGroup>-->
```
Comment (zuren) [inline LBHTenancyAPI/LBHTenancyAPI.csproj]: Should we just remove this config outright, rather than commenting it out? Looks like a security concern to have service config in here.

File: LBHTenancyAPITest/Connected Services/ArrearsAgreementService/ConnectedService.json [hunk 10]
```diff
@@ -0,0 +1,24 @@
+﻿{
+  "ProviderId": "Microsoft.VisualStudio.ConnectedService.Wcf",
+  "Version": "15.0.20119.1312",
+  "ExtendedData": {
+    "Uri": "http://lbhuhwwebp01/civica-uh-gateway-dev/webservices/ArrearsAgreementService.svc",
```
Comment (zuren) [inline LBHTenancyAPITest/Connected Services/ArrearsAgreementService/ConnectedService.json]: Same here r.e. security.

File: LBHTenancyAPITest/Dockerfile [hunk 11]
```diff
@@ -1,12 +1,12 @@
...
 WORKDIR /app
 EXPOSE 80
 
-FROM microsoft/aspnetcore-build:2.0 AS build
+FROM microsoft/dotnet:latest AS build
 WORKDIR /src
 COPY LBHTenancyAPI.sln ./
-COPY LBHTenancyAPITest/LBHTenancyAPITest.csproj LBHTenancyAPITest/
-RUN dotnet restore -nowarn:msb3202,nu1503
+COPY LBHTenancyAPITest/LBHTenancyAPITest.csproj LBHTenancyAPITest/
```
Comment (zuren) [inline LBHTenancyAPITest/Dockerfile]: Same here - necessary change?

File: LBHTenancyAPITest/Helpers/Fake.cs [hunk 12]
```diff
@@ -122,5 +123,42 @@ public static ArrearsActionDiaryEntry GenerateActionDiary()
...
+
+            return new ArrearsActionResponse
+            {
+                ArrearsAction = new ArrearsActionLogDto
+                {
+                    TenancyAgreementRef = request.ArrearsAction.TenancyAgreementRef,
+                    ActionBalance = random.Finance.Amount(),
+                    ActionCode = random.Random.Hash(3),
+                    ActionCategory = random.Random.Hash(10),
+                    //Comment = random.Random.Hash(100)
```
Comment (zuren) [inline LBHTenancyAPITest/Helpers/Fake.cs]: Why is this commented out?

File: LBHTenancyAPITest/Helpers/Fake.cs [hunk 13]
```diff
@@ -122,5 +123,42 @@ public static ArrearsActionDiaryEntry GenerateActionDiary()
...
+                },
+                ErrorCode = 100,
```
Comment (zuren) [inline LBHTenancyAPITest/Helpers/Fake.cs]: ErrorCode 100?

File: LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+                    ActionCode = "GEN",
+                    Comment = "Testing",
+                    TenancyAgreementRef = "000017/01"
+                },
+                DirectUser = new UserCredential
+                {
+                    UserName = "HackneyAPI",
+                    UserPassword = "Hackney1"
+                },
+                SourceSystem = "HackneyAPI"
```
Comment (zuren) [inline LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs]: Are these real credentials?

File: LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json (hunk 5 above)
Comment (zuren) [reply LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json]: Nah, it's an internal host.

**Comment by zuren** [comment]
Should probably rebase these commits down to something more manageable/meaningful. Loads of noise from the previous commit history.

File: LBHTenancyAPI.v3.ncrunchsolution (hunk 4 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI.v3.ncrunchsolution]: Yep we can gitignore this

File: LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json (hunk 5 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Connected Services/AgreementService/ConnectedService.json]: Yep, that's a good call we need to strip that out

File: LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs (hunk 6 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: The using statements or comment?

File: LBHTenancyAPI/LBHTenancyAPI.csproj (hunk 9 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/LBHTenancyAPI.csproj]: Agreed

File: LBHTenancyAPI.sln (hunk 3 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI.sln]: This is automatically done by Visual Studio, we could remove it but it would most likely get re-added by Visual Studio. It's subject to churn but Visual Studio usually manages it.

File: LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs (hunk 6 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: That's autogenerated, I'll remove it.

File: LBHTenancyAPI/Controllers/TenanciesController.cs (hunk 7 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Controllers/TenanciesController.cs]: Yep, I'll reverse

File: LBHTenancyAPITest/Connected Services/ArrearsAgreementService/ConnectedService.json (hunk 10 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPITest/Connected Services/ArrearsAgreementService/ConnectedService.json]: Yep agreed

File: LBHTenancyAPITest/Dockerfile (hunk 11 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPITest/Dockerfile]: removed

File: LBHTenancyAPITest/Helpers/Fake.cs (hunk 12 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPITest/Helpers/Fake.cs]: Unsure, pulled in from previous branch, checked and there is no comment field. Removed

File: LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs (hunk 14 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs]: Possibly, I'll remove now

File: LBHTenancyAPI/Extensions/Validation/CreateArrearsActionDiaryExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        }
+    }
+
+    public class ArrearsActionCreateRequestValidator : AbstractValidator<ArrearsActionCreateRequest>
+    {
+        public ArrearsActionCreateRequestValidator()
+        {
+            RuleFor(x => x).NotNull();
+            RuleFor(x => x.ArrearsAction).NotNull();
+            RuleFor(x => x.ArrearsAction.TenancyAgreementRef).NotEmpty().WithMessage("Please specify a first name");
```
[the remark this answers is not in the document]
Comment (StevenLeighton21) [reply LBHTenancyAPI/Extensions/Validation/CreateArrearsActionDiaryExtensions.cs]: I probably wouldn't consider it a priority, but we could add a useful message to each field in that case

File: LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs [hunk 16]
```diff
@@ -0,0 +1,31 @@
+using System;
+using System.Threading.Tasks;
+using AgreementService;
+using LBHTenancyAPI.Interfaces;
+using LBHTenancyAPI.UseCases;
+
+namespace LBHTenancyAPI.Gateways
+{
+    /// <summary>
```
Comment (StevenLeighton21) [inline LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs]: I'd probably omit these unless they're contributing to some doc generation we're using. We aren'y currently doing any generation so would remove them for consistency.

File: LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+            var request = _builder.BuildArrearsRequest(new ArrearsActionCreateRequest
+            {
+                ArrearsAction = new ArrearsActionInfo()
+            });
+            //assert
+            Assert.IsType<ArrearsActionCreateRequest>(request);
+        }
+
+        [Fact]
+        public void WhenGivenActionDiaryDetails_BuildValidArrearsActionDiaryRequest()
```
Comment (StevenLeighton21) [inline LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs]: I think this test would benefit from triangulation, either by randomising the input or by providing a second example test with different input

File: LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            //pre-arrange
+            var credentialsService = new Mock<ICredentialsService>();
+            _builder = new ArrearsServiceRequestBuilder(credentialsService.Object);
+        }
+
+        [Fact]
+        public void Return_A_Built_Request_Object()
+        {
+            //arrange
```
Comment (StevenLeighton21) [inline LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs]: Suggest removing these comments. Personal preference :)

File: LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+                SourceSystem = "TestSystem"
+            };
+
+            fakeArrearsAgreementService.Setup(s => s.CreateArrearsActionAsync(It.IsAny<ArrearsActionCreateRequest>()))
+                .ReturnsAsync(Fake.CreateArrearsActionAsync(request));
+
+            //act
+            var response = await classUnderTest.CreateActionDiaryEntryAsync(request);
+            //assert
+            Assert.Equal(response.ArrearsAction.TenancyAgreementRef, request.ArrearsAction.TenancyAgreementRef);
```
Comment (StevenLeighton21) [inline LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs]: Suggest these tests could assert against each field in the request, and triangulate the test with a second example/randomised input

File: LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        private readonly ICreateArrearsActionDiaryUseCase _createArrearsActionDiaryUseCase;
+
+        public ArrearsActionDiaryController (ICreateArrearsActionDiaryUseCase createArrearsActionDiaryUseCase)
+        {
+            _createArrearsActionDiaryUseCase = createArrearsActionDiaryUseCase;
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Post([FromBody][Required] ArrearsActionCreateRequest request)
```
Comment (StevenLeighton21) [inline LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: Sorry if this is an obvious question, I don't quite understand how this endpoint receives the info/creates the ArrearsActionCreatRequest. How does an outside consumer use this API?

File: LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs (hunk 18 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs]: This will hopefully serve to help the team build good practise in their approach to TDD

File: LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs (hunk 16 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Gateways/ArrearsActionDiaryGateway.cs]: I try to comment where ever possible, especially as a contractor, my time on the project is limited and eventually someone will have to make sense of my work.

File: LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs (hunk 20 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: The swagger docs should help out in this regard. /swagger. But essentially you pass a json object in as a body in a HTTP POST request, with an ContentType of "application/json"

File: LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs (hunk 17 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPITest/Test/Services/ArrearsServiceRequestBuilderTests.cs]: Ok yep, I'll try and do that

File: LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs (hunk 19 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPITest/Test/Gateways/ArrearsActions/UHArrearsDiaryCreationGatewayTest.cs]: Yep, I'll try to do that.

**Comment by madetech-jeffpinkham** [comment]
@StevenLeighton21 and @zuren I've actioned the test suggestions, and I'll defer the validation piece to another PR as that's another piece. For the moment I'll hand hold with any API requests until I get that in.

File: LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using AgreementService;
+using LBHTenancyAPI.Extensions.Validation;
+using LBHTenancyAPI.UseCases.ArrearsActions;
+
+namespace LBHTenancyAPI.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/v1/tenancies/arrearsactiondiary/")]
```
Comment (zuren) [inline LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: Suggest we change to `/api/v1/tenancies/arrears-action-diary` to be more consistent.

Question around whether we should stick to the "arrears action diary" term, or call this something like `/api/v1/tenancies/arrears_event` ?

Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: Yeah I've been toying with the naming of it the official web service method is called CreateArrearsActionAsync

Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: Arrears actions Controller?

**Comment by madetech-jeffpinkham** [comment]
Ok sounds good :)

File: LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs (hunk 21 above)
Comment (madetech-jeffpinkham) [reply LBHTenancyAPI/Controllers/ArrearsActionDiaryController.cs]: made it /api/v1/tenancies/arrears-action-diary
